primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,184 @@
+		let res = self.value_ref.mutate(
+			|s: Option<Option<Timestamp>>| -> Result<Timestamp, Option<Timestamp>> {
+				match s {
+					// no lock set, we can safely acquire it
+					None => Ok(expires_at),
+					// lock is set, but it's old. We can re-acquire it.
+					Some(Some(current_good_until)) if current_good_until < now => Ok(expires_at),
+					// lock is present and is still active
+					Some(Some(current_good_until)) => Err(Some(current_good_until)),
+					_ => Err(None),
+				}
+			},
+		);
+		match res {
+			Ok(Ok(_)) => {
+				self.locked_until = Some(expires_at);
+				Ok(())
+			}
+			Ok(Err(timestamp)) => Err(Some(timestamp)), // failed to set the new value, but could read the current
+			Err(e) => Err(e),                           // forward the remaining value
+		}
+	}
+
+	/// Attempt to lock the storage entry.
+	///
+	/// Returns a lock guard on success, otherwise an error containing `None` in
+	/// case the mutex was already unlocked before, or if the lock is still held
+	/// by another process `Err(Some(expiration_timestamp))`.
+	pub fn try_lock<'b>(&'b mut self) -> Result<StorageLockGuard<'a, 'b>, Option<Timestamp>>
+	where
+		'a: 'b,
+	{
+		if self.locked_until.is_none() {
+			match self.try_lock_inner(Duration::from_millis(STORAGE_LOCK_DEFAULT_EXPIREY_DURATION))
+			{
+				Ok(_) => Ok(StorageLockGuard::<'a, 'b> { lock: Some(self) }),
+				Err(e) => Err(e),
+			}
+		} else {
+			Err(self.locked_until)
+		}
+	}
+
+	/// Try grabbing the lock until its expiry is reached.
+	///
+	/// Returns an error if the lock expired before it could be cought
+	pub fn spin_lock<'b, 'c>(&'b mut self) -> Result<StorageLockGuard<'a, 'c>, ()>
+	where
+		'a: 'b,
+		'b: 'c,
+	{
+		if self.locked_until.is_none() {
+			loop {
+				// blind attempt on locking
+				let expires_at = match self
+					.try_lock_inner(Duration::from_millis(STORAGE_LOCK_DEFAULT_EXPIREY_DURATION))
+				{
+					Ok(_) => return Ok(StorageLockGuard::<'a, 'b> { lock: Some(self) }),
+					Err(None) => return Err(()),
+					Err(Some(expires_at)) => expires_at,
+				};
+				let remainder: Duration = offchain::timestamp().diff(&expires_at);
+				// do not snooze the full duration, but instead snooze max 100ms
+				// it might get unlocked in another thread
+				// consider adding some additive jitter here
+				let snooze =
+					core::cmp::min(remainder.millis(), STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE);
+				sp_io::offchain::sleep_until(
+					offchain::timestamp().add(Duration::from_millis(snooze)),
+				);
+			}
+		} else {
+			Err(())
+		}
+	}
+
+	/// Explicitly unlock the lock.
+	///
+	/// Does nothing if the lock was alrady unlocked before.
+	pub fn unlock(&mut self) {
+		if let Some(_) = self.locked_until.take() {
+			self.value_ref.remove();
+		}
+	}
+}
+
+/// RAII style guard for a lock.
+pub struct StorageLockGuard<'a, 'b> {
+	lock: Option<&'b mut StorageLock<'a>>,
+}
+
+impl<'a, 'b> StorageLockGuard<'a, 'b> {
+	/// Consume the guard and unlock the underlying lock.
```

Comment:
```suggestion
	/// Consume the guard but DON'T unlock the underlying lock.
	/// 
	/// This can be used to finish off-chain worker execution while keeping the lock for it's desired expiry time.
```